Feature/subscribe

danhanson opened the pull request:
Adam broke my branch #24, so here is the pull request for pushes
When you create a vehicle, you are subscribed to receive route updates for that vehicle

danhanson committed:
added has parent and made subscription management classes

danhanson committed:
added Router object

danhanson committed:
support pushes for creation of commodities and vehicles

aj-michael commented on app/io/pathfinder/routing/Router.scala:
```diff
@@ -0,0 +1,105 @@
+        recalculate()
+    }
+
+    // Dan wrote an infinite loop here.
```
Best part! :+1:

aj-michael commented on conf/evolutions/default/1.sql:
```diff
@@ -15,6 +15,7 @@ create table commodity (
   endLatitude               double not null,
```
Please remove this file.

aj-michael commented on app/io/pathfinder/data/EbeanCrudDao.scala:
```diff
@@ -4,7 +4,7 @@ import com.avaje.ebean.Model
 import scala.collection.JavaConversions.asScalaBuffer
 
-abstract class EbeanCrudDao[K,M <: Model](finder: Model.Find[K,M]) extends CrudDao[K,M] {
+class EbeanCrudDao[K,M <: Model](finder: Model.Find[K,M]) extends CrudDao[K,M] {
```
Please add a comment explaining the purpose of this class.

aj-michael commented on app/io/pathfinder/data/ObserverDao.scala:
```diff
@@ -0,0 +1,49 @@
+
+/**
+ * subclasses can use callbacks to listen to changes to models in the specified dao
+ * @param dao
```
For these `@param` tags, please either remove them or add descriptions. Also, please be a little more concrete in your description of the class. I cannot explain what it means from this.

aj-michael commented on app/io/pathfinder/websockets/WebSocketActor.scala:
```diff
@@ -1,9 +1,15 @@
+import scala.concurrent.ExecutionContext.Implicits.global
+import play.api.libs.iteratee.{Iteratee, Concurrent}
 
 object WebSocketActor {
```
Why is this an object and not a class? If all of our web socket requests are routed through a singleton, we are limiting our scaling potential

aj-michael commented on app/io/pathfinder/websockets/WebSocketActor.scala:
```diff
@@ -24,18 +30,48 @@ class WebSocketActor (
     controllers: Map[ModelType, WebSocketController]
 ) extends Actor {
     import WebSocketMessage._
+    Router
```
This line should not go here. If we put it here, the Router object is not initialized until the first socket request is made. This causes the first socket connection to be delayed.

aj-michael commented on app/io/pathfinder/websockets/WebSocketActor.scala:
```diff
@@ -1,9 +1,15 @@
+import io.pathfinder.models.{Vehicle, Commodity}
+import io.pathfinder.routing.Router
+import io.pathfinder.websockets.ModelTypes.ModelType
+import io.pathfinder.websockets.controllers.VehicleSocketController._
```
You know my feelings about wildcard imports.

aj-michael commented on app/io/pathfinder/models/Commodity.scala:
```diff
@@ -5,11 +5,12 @@ import javax.persistence.{Column, Entity, GeneratedValue, GenerationType, Id, Ma
 object Commodity {
     val finder: Model.Find[Long,Commodity] = new Model.Finder[Long,Commodity](classOf[Commodity])
-    object Dao extends EbeanCrudDao[Long,Commodity](finder)
+    object Dao extends WebSocketDao[Commodity](finder)
```
Why is this dao specific to web sockets? I don't think that the data layer should be coupled to the transport layer.

danhanson replied:
It is for push requests, the websocket dao is intended to push notifications to subscribed websockets. Anyone using the dao to access its data doesn't need to worry about that though

aj-michael committed:
Moved existing test files to proper packages.

We should put the test files in the same package as the source code they are testing, just under /test instead of /app. This allows us to not have to import the classes we are testing and gives us access to package private fields.

aj-michael committed:
Removed unused Scala file.

aj-michael committed:
Renamed test files for consistency.

aj-michael committed:
Added test for Vehicle model.

aj-michael committed:
Serialization/deserialization tests for Commodities.

danhanson committed:
removed evelotions/default/1.sql

danhanson committed:
Added tests for Route writing and renamed Route.format to Route.format

Added testing for route writing to ensure that the routes are written as
json to increase test coverage.

danhanson committed:
Merge branch 'feature/subscribe' of https://github.com/CSSE497/pathfinder-server into feature/subscribe

I think I removed the 1.sql evolution file because it alway conflicts
and doesn't do anything

aj-michael commented:
LGTM
